spring-kafka/src/main/java/org/springframework/kafka/listener/DeadLetterPublishingRecoverer.java

```diff
@@ -335,39 +434,68 @@ protected void publish(ProducerRecord<Object, Object> outRecord, KafkaOperations
 		catch (Exception e) {
 			this.logger.error(e, () -> "Dead-letter publication failed for: " + outRecord);
 		}
+		if (this.failIfSendResultIsError) {
+			verifySendResult(outRecord, sendResult);
+		}
+	}
+
+	private void verifySendResult(ProducerRecord<Object, Object> outRecord,
+								@Nullable ListenableFuture<SendResult<Object, Object>> sendResult) {
+		if (sendResult == null) {
+			throw new KafkaException("Dead-letter publication failed for: " + outRecord);
+		}
+		try {
+			sendResult.get(this.waitForSendResultTimeout, TimeUnit.MILLISECONDS);
+		}
+		catch (InterruptedException | ExecutionException | TimeoutException e) {
```

Comment:
Interrupt needs to be in a separate catch class with `Thread.currentThread.interrupt()` before throwing; otherwise we "eat" the interrupt and some downstream interruptible will block.